appengine/genetics/js/blocks.js

```diff
@@ -0,0 +1,463 @@
+/**
+ * Blockly Games: Genetics Blocks
+ *
+ * Copyright 2016 Google Inc.
+ * https://github.com/google/blockly-games
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Blocks for Blockly's Genetics application.
+ * @author kozbial@google.com (Monica Kozbial)
+ */
+'use strict';
+
+goog.provide('Genetics.Blocks');
+
+goog.require('Blockly');
+goog.require('Blockly.JavaScript');
+goog.require('BlocklyGames');
+goog.require('BlocklyGames.JSBlocks');
+
+/**
+ * HSV hue for all genetics mouse function blocks.
+ */
+Genetics.Blocks.GENETICS_MOUSEFUNCTIONS_HUE = 360;
+/**
+ * HSV hue for all genetics blocks (that are not mouse functions).
+ */
+Genetics.Blocks.GENETICS_HUE = 20;
+
+// Extensions to Blockly's language and JavaScript generator.
+
+/**
+ * Generates the JavaScript from a "mouse function" block (pickFight,
+ * chooseMate, mateAnswer) given the name and arguments of the function.
+ * @param {string} funcName The name of the mouse function.
+ * @param {string} args A comma separated string of the argument variable names.
+ * @param {!Blockly.Block} block The block.
+ * @private
+ */
+Blockly.JavaScript['genetics_generateMouseFunctionJS_'] = function(funcName,
```

Comment:
Per our chat, it looks like this is based on the procedures_defreturn block, rather than the other "normal" blocks which return a code string. However, I believe that block is a special case, and is written like it is in order to allow users to create functions with arbitrary names, which isn't really a consideration here.

So, I feel a bit uncomfortable about this. Here's why: I see Blockly Games as kind of an extension to Blockly, so IMO it should use Blockly's public APIs. However, you have a function in here that reaches into Blockly.JavaScript.definitions_ and modifies that directly. Also, I don't think there's precedent for doing this in the other games. On the other hand, I realize that going the more "normal" route is likely to result in some code duplication, so it's a question of trade-off (though I believe this consideration applies in the other games, too).

In any case, per our chat, I'm leaving this comment here so that you can resolve it when you've made a decision, now that you're aware of the above.